cmd/create/idp/htpasswd.go

```diff
@@ -187,19 +187,39 @@ func usernameValidator(val interface{}) error {
 
 func passwordValidator(val interface{}) error {
 	if password, ok := val.(string); ok {
-		notAsciiOnly, _ := regexp.MatchString(`[^\x20-\x7E]`, password)
+		re := regexp.MustCompile(`[^\x20-\x7E]`)
+		invalidChars := re.FindAllString(password, -1)
+		notAsciiOnly := len(invalidChars) > 0
 		containsSpace := strings.Contains(password, " ")
 		tooShort := len(password) < 14
-		if notAsciiOnly || containsSpace || tooShort {
-			return fmt.Errorf(
-				"password must be at least 14 characters (ASCII-standard) without whitespaces")
+		pwdErrors := []string{}
+		if notAsciiOnly {
+			pwdErrors = append(pwdErrors, fmt.Sprintf("must not contain special characters [%s]",
+				strings.Join(invalidChars, ", ")))
+		}
+		if containsSpace {
+			pwdErrors = append(pwdErrors, "must not contain whitespace")
+		}
+		if tooShort {
+			pwdErrors = append(pwdErrors, fmt.Sprintf("must be at least 14 characters (got %d)", len(password)))
+		}
+		if len(pwdErrors) == 1 {
+			return fmt.Errorf("Password " + pwdErrors[0])
+		} else if len(pwdErrors) > 1 {
+			msg := "Password "
+			for i := 0; i < len(pwdErrors)-1; i++ {
+				msg += pwdErrors[i] + ", "
```

Comment:
Would it be possible to use strings.join instead and possibly drop the condition?

Reply:
@gdbranco thanks for the review! I'd like to keep the condition since I think having the "and" before the last list item makes the message a bit more readable, but I switched the code to use `strings.Join` to make it cleaner.